### PR TITLE
9142 support dashes and uuids as table names

### PR DESCRIPTION
**Pull request by juanignaciosl**
This fixes (most of) #9142. Checked:

- Create map based on table (Editor and Builder).
- Open table (Editor and Builder).

Other problems might arise but I think that this fixes original user issues and its a safe, small first step.

### CODE REVIEW

**Comment by juanignaciosl** [comment]
@javitonino CR backend, please.
@xavijam CR frontend, please.

File: app/controllers/admin/visualizations_controller.rb [hunk 1]
```diff
@@ -630,11 +630,15 @@ def get_visualization_and_table(table_id, schema, filter)
     user = Carto::User.where(username: schema).first
     # INFO: organization public visualizations
     if user
-      visualization = get_priority_visualization(table_id, user_id: user.id)
+      visualization = get_priority_visualization(table_id, user_id: user.id, force_name: false)
+      visualization = get_priority_visualization(table_id, user_id: user.id, force_name: true) unless visualization
```
Comment (javitonino) [inline app/controllers/admin/visualizations_controller.rb]: Maybe this logic can be moved inside `get_priority_visualization`?

File: lib/carto/table_utils.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    end
+
+    def safe_schema_name_quoting(schema_name)
+      dashes_quoting(schema_name)
+    end
+
+    private
+
+    def dashes_quoting(name)
+      name && name.include?('-') ? "\"#{name}\"" : name
```
Comment (javitonino) [inline lib/carto/table_utils.rb]: I think checking for quotes here is worth it, to avoid double quoting.

Also, it would be nice to quote the whole spectrum of postgresql identifiers, not just hyphens:
https://www.postgresql.org/docs/9.3/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
So anything that it is not `[a-z][A-Z][0-9]_$` gets quotes. Anyway, that may be going too far, and can be easily added later, so maybe better left for later.

File: app/models/carto/layer.rb [hunk 3]
```diff
@@ -167,7 +170,7 @@ def default_query(user = nil)
         table_name = sym_options[:table_name]
 
         if table_name.present? && !table_name.include?('.') && user_name.present? && user_username != user_name
-          %{ select * from "#{user_name}".#{table_name} }
+          %{ select * from "#{user_name}"."#{table_name}" }
```
Comment (javitonino) [inline app/models/carto/layer.rb]: Why not `safe_table_name_quoting` here?

File: app/models/layer/presenter.rb [hunk 4]
```diff
@@ -91,9 +94,9 @@ def to_poro
           # if the table_name already have a schema don't add another one
           # this case happens when you share a layer already shared with you
           if user_name != options[:viewer_user].username && !poro['options']['table_name'].include?('.')
-            poro['options']['table_name'] = schema_name.include?('-') ?
-              "\"#{schema_name}\".#{poro['options']['table_name']}" :
-              "#{schema_name}.#{poro['options']['table_name']}"
+            safe_schema = safe_schema_name_quoting(schema_name)
+            safe_table_name = safe_table_name_quoting(poro['options']['table_name'])
+            poro['options']['table_name'] = "#{safe_schema}.#{safe_table_name}"
```
Comment (javitonino) [inline app/models/layer/presenter.rb]: If we move the logic to avoid double quoting to TableUtils, this can become cleaner. Also, this only quotes the table name if the user must be quoted too, right?

**Comment by javitonino** [comment]
Left some comments. I think we can extract some of the logic to the new helpers in order to simplify the code. Let me know what are your thoughts on that!

File: lib/assets/javascripts/cartodb/models/models_utils.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+(function () {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/models/models_utils.js]: I feel you were looking for [this file](https://github.com/CartoDB/cartodb/blob/master/lib/assets/javascripts/cartodb/old_common/utils.js) for the old editor.

File: lib/build/files/js_files.js [hunk 6]
```diff
@@ -136,6 +136,7 @@ module.exports = files = {
   ],
 
   models: [
+    'lib/assets/javascripts/cartodb/models/models_utils.js',
```
Comment (xavijam) [inline lib/build/files/js_files.js]: Not needed, check above :)

**Comment by xavijam** [comment]
Code looks great but be careful with the comment about the unnecessary `models_utils` file.

File: app/controllers/admin/visualizations_controller.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/controllers/admin/visualizations_controller.rb]: Makes sense 👍

File: app/models/carto/layer.rb (hunk 3 above)
Comment (juanignaciosl) [inline app/models/carto/layer.rb]: I overlooked this, thanks!

File: app/models/layer/presenter.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/models/layer/presenter.rb]: More precisely, this only quotes the table name _if the user schema needs to be added_ (no matter about its quoting needs). This might not quote some cases, but I haven't found any. The `!poro['options']['table_name'].include?('.')` make me think that a full schema + table name string is passed here, and I prefer not adding _more_ logic. I'll move the assignment to `TableUtils`, though.

File: lib/assets/javascripts/cartodb/models/models_utils.js (hunk 5 above)
Comment (juanignaciosl) [inline lib/assets/javascripts/cartodb/models/models_utils.js]: Yup. On it, thanks! cc @viddo

File: lib/carto/table_utils.rb (hunk 2 above)
Comment (juanignaciosl) [inline lib/carto/table_utils.rb]: Ok to the first. To the second, as you said, I prefer keeping the scope of this PR as narrow as possible.

**Comment by juanignaciosl** [comment]
@javitonino , @xavijam next CR after your comments, please.

File: lib/carto/table_utils.rb [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    def safe_schema_and_table_quoting(schema_name, table_name)
+      safe_schema = safe_schema_name_quoting(schema_name)
+      safe_table_name = safe_table_name_quoting(table_name)
+      "#{safe_schema}.#{safe_table_name}"
+    end
+
+    private
+
+    def dashes_quoting(name)
+      name && name.include?('-') && !name.match(/".*"/) ? "\"#{name}\"" : name
```
Comment (javitonino) [inline lib/carto/table_utils.rb]: `name.match` in Ruby doesn't actually match in the traditional sense (full string match) but actually searches for a match in the string :disappointed:. To get actual match behaviour you can use start/end of line characters in the regex: `name.match(/^".*"$/)`, this will only match the quotes at the start/end of the string.

I am not sure at all about this change, maybe the former behaviour is better. Just wanted to warn you about this behaviour.

File: app/controllers/visualizations_controller_helper.rb [hunk 8]
```diff
@@ -103,4 +110,23 @@ def generate_anonymous_map_vizjson3(visualization, params)
     Carto::Api::VizJSON3Presenter.new(visualization)
                                  .to_anonymous_map_vizjson(https_request: is_https?, vector: params[:vector] == 'true')
   end
+
+  private
```
Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: Does this work? (This is a module, not a class)

**Comment by javitonino** [comment]
Left a couple of minor comments, but LGTM. :heart: to see the quoting logic extracted to a single place.

File: lib/carto/table_utils.rb (hunk 7 above)
Comment (juanignaciosl) [inline lib/carto/table_utils.rb]: Oh, true, I forgot about the needed pattern (I've already seen this in the past). In fact, [to match _string_ beginning and end you should use `\A` and \Z`](http://ruby-doc.org/core-2.2.0/Regexp.html#class-Regexp-label-Anchors). Fixing.

File: app/controllers/visualizations_controller_helper.rb (hunk 8 above)
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: Yes:

```ruby
irb(main):001:0> class XXX
irb(main):002:1> include Carto::TableUtils
irb(main):003:1> end
=> XXX
irb(main):004:0> x = XXX.new
=> #<XXX:0x00000002d28c08>
irb(main):005:0> x.dashes_quoting('x')
NoMethodError: private method `dashes_quoting' called for #<XXX:0x00000002d28c08>
```

Module _class_ methods are problematic, module _instance_ methods are not.

Comment (javitonino) [inline app/controllers/visualizations_controller_helper.rb]: :+1: Thanks for the detailed explanation!

**Comment by juanignaciosl** [comment]
@javitonino next 👀 , sorry. I wanted to test the regexp fix quicky so I did a test that led me to use a broader regexp (the one that you suggested, indeed), as I couldn't actually test it otherwise.

**Comment by javitonino** [comment]
Looks :+1: Be careful with the latest merge though, as I just reverted the layer dependencies patch.

**Comment by juanignaciosl** [comment]
Oh, sh*t. It had conflicts indeed. Reverting merge and crossing fingers...

**Comment by juanignaciosl** [comment]
Everything looks ok after the (re)merge, I'll merge this tomorrow morning unless @xavijam or @javitonino says something.

**Comment by juanignaciosl** [comment]
@gfiorav , could you CR this, please? @javitonino did the first CR but I've changed some details around table registration because of ghost tables and maybe you see something that I don't. 

@javitonino and @xavijam , take a look as well if you want. It's deployed in ded01 in case that you want to try to break it 🔨

**Comment by javitonino** [comment]
Made it go boom with an `@` in the name. I just thought that this case may be supported (as backend was changed to allow for it), but frontend still only detects hyphens.

I also ran into a problem running analyses, they never got marked as finished (although I could see the results). Most likely not related to this PR at all.

Overall, it seems everything works, so nothing left but a congratulation on a job well done :clap:

File: app/controllers/visualizations_controller_helper.rb [hunk 9]
```diff
@@ -103,4 +111,23 @@ def generate_anonymous_map_vizjson3(visualization, params)
...
+
+  def get_priority_visualization_forcing_name(visualization_id, force_name: false, user_id: nil, organization_id: nil)
```
Comment (gfiorav) [inline app/controllers/visualizations_controller_helper.rb]: minor: this method name doesn't sound like it will optionally force a name :trollface:

File: app/helpers/bounding_box_helper.rb [hunk 10]
```diff
@@ -1,6 +1,8 @@
 # encoding: utf-8
+require_relative '../../lib/carto/table_utils'
```
Comment (gfiorav) [inline app/helpers/bounding_box_helper.rb]: why not `require_dependency 'carto/table_utils'?

File: app/models/table.rb [hunk 11]
```diff
@@ -664,7 +665,7 @@ def name=(value)
     value = value.downcase if value
     return if value == @user_table[:name] || value.blank?
 
-    new_name = get_valid_name(value)
+    new_name = register_table_only ? value : get_valid_name(value)
```
Comment (gfiorav) [inline app/models/table.rb]: Why this logic change?

File: app/helpers/bounding_box_helper.rb (hunk 10 above)
Comment (juanignaciosl) [inline app/helpers/bounding_box_helper.rb]: I don't know the precise details but `require_dependency` (my default) crashed. I think that it's caused by Rails loading order. This is inside `helpers` (in spite of not being a view helper) and it might be the cause. I promise that I tried :_)

File: app/controllers/visualizations_controller_helper.rb (hunk 9 above)
Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: I know, but I couldn't come up with a better one. I'll accept any better suggestion :-)

File: app/views/admin/visualizations/embed_map.html.erb [hunk 12]
```diff
@@ -1,3 +1,5 @@
+<% require_dependency 'static_maps_url_helper' %>
```
Comment (gfiorav) [inline app/views/admin/visualizations/embed_map.html.erb]: What does this have to do with anything :/

File: app/models/table.rb (hunk 11 above)
Comment (juanignaciosl) [inline app/models/table.rb]: This is the most tricky part (and the one why I summoned you). We don't want dashes everywhere now, we just want to support it when the user has forced them (SQL API, almost exclusively). The problem is that this logic at `name` setter provokes the rewrite. `def name=` would change dashes to underscores. This breaks ghost tables for tables with dashes. The best fix I've come up with is not changing it if "you're only registering" (an existing one). That's why [I've changed the order of that set at propagation](https://github.com/CartoDB/cartodb/pull/9145/files/57bb7a45d49965834b793632be28a29dd1177ffd#diff-39a221071cc72b160782078c80b842bbR833).

Comment (juanignaciosl) [inline app/models/table.rb]: PS: this can be risky but currently needed if you don't want to propagate a patch flag :(. But please let me know if you think that this can break something. As 👻  expert I'd love that you break it 😅

File: lib/carto/table_utils.rb [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+    def can_be_quoted?(name)
+      name && !name.include?('.')
+    end
+  end
+
+  # Used for class methods
+  class TableUtilsInstance
+    include TableUtils
+  end
```
Comment (gfiorav) [inline lib/carto/table_utils.rb]: I'm sure there is some pattern I don't know about, but from total ignorance:

Why not use class methods or include the module?
 
Like
```ruby
Carto::TableUtils.safe_table_name_quoting('bananas')
```

or
```ruby
include Carto::TableUtils

safe_table_name_quoting('bananas')
```

The instance thing _looks_ a bit hacky and is less understandable IMHO.

Super minor, of course.

File: app/views/admin/visualizations/embed_map.html.erb (hunk 12 above)
Comment (juanignaciosl) [inline app/views/admin/visualizations/embed_map.html.erb]: If fixes an eventual symbol error because of Rails hot deploy.

**Comment by gfiorav** [comment]
Nice work and nice tests!

I've made a few minor comments and generally I think we should revise mixing https://github.com/CartoDB/cartodb/blob/master/lib/carto/db/sanitize.rb and https://github.com/CartoDB/cartodb/blob/master/lib/carto/valid_table_name_proposer.rb into this or the other way around.

📦 it when ready!

File: app/controllers/visualizations_controller_helper.rb (hunk 9 above)
Comment (gfiorav) [inline app/controllers/visualizations_controller_helper.rb]: I think the reeky thing to do would be defining both `get_priority_visualization_forcing_name` and `get_priority_visualization` and make the decision based on force_name outside. That way you avoid the control parameter stuff... 

I wouldn't do it though 😆

Comment (juanignaciosl) [inline app/controllers/visualizations_controller_helper.rb]: The problem is that `load_visualization_from_id_or_name_guessing` also needs the parametrized version of the method... I'll keep it this way.

File: lib/carto/table_utils.rb (hunk 13 above)
Comment (juanignaciosl) [inline lib/carto/table_utils.rb]: Oh. My problem was that I needed it at `BoundingBoxHelper` module, which is a utils module with _class_ methods. In that case `include` doesn't work, as it'd include it as _instance_ methods. Nevertheless I forgot that [`extend` can be used in this situation](http://www.railstips.org/blog/archives/2009/05/15/include-vs-extend-in-ruby/).

**Comment by juanignaciosl** [comment]
@gfiorav I took into account your comment, last CR before :shipit: !

**Comment by javitonino** [comment]
Looking at other stuff and found a minor UI glitch that I think also affects this PR. If you see in the layer listing, you may see `SQL` in the source analysis, which is used to indicate that a custom SQL query is being used. I think the frontend detects this by looking at the query and matching it to "SELECT * FROM <table_name>". It may fail due to adding the username/quotes. Up to you if you want to include it in this PR (which is already too big) or open a ticket for another one, as this is minor anyway.

**Comment by xavijam** [comment]
We can manage that problem in other PR, let's try to keep going with this one please.

**Comment by gfiorav** [comment]
Looking sharp! ✏️ 

📦 it!

**Comment by juanignaciosl** [comment]
Ok, let's do it.
